aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -419,10 +410,278 @@ impl AptosVM {
+                storage,
+                session,
+                gas_meter,
+                txn_data,
+                cleanup_args,
+            )?
+        };
+
+        // TODO(Gas): Charge for aggregator writes
+        self.success_transaction_cleanup(
+            storage,
+            final_change_set_ext,
+            gas_meter,
+            txn_data,
+            log_context,
+        )
+    }
+
+    fn execute_multisig_entry_function<SS: MoveResolverExt>(
+        &self,
+        session: &mut SessionExt<SS>,
+        gas_meter: &mut AptosGasMeter,
+        multisig_address: AccountAddress,
+        payload: &EntryFunction,
+        new_published_modules_loaded: &mut bool,
+    ) -> Result<(), VMStatus> {
+        let function =
+            session.load_function(payload.module(), payload.function(), payload.ty_args())?;
+        // This transaction is now being executed as the multisig account.
+        // If txn args are not valid, we'd still consider the multisig transaction as executed but
+        // failed. This is primarily because it's unrecoverable at this point.
+        let args = verifier::transaction_arg_validation::validate_combine_signer_and_txn_args(
+            session,
+            vec![multisig_address],
+            payload.args().to_vec(),
+            &function,
+        )?;
+        session
+            .execute_entry_function(
+                payload.module(),
+                payload.function(),
+                payload.ty_args().to_vec(),
+                args,
+                gas_meter,
+            )
+            .map_err(|e| e.into_vm_status())?;
+
+        // Resolve any pending module publishes in case the multisig transaction is deploying
+        // modules.
+        self.resolve_pending_code_publish(session, gas_meter, new_published_modules_loaded)?;
+        Ok(())
+    }
+
+    fn success_multisig_payload_cleanup<S: MoveResolverExt + StateView, SS: MoveResolverExt>(
+        &self,
+        storage: &S,
+        session: SessionExt<SS>,
+        gas_meter: &mut AptosGasMeter,
+        txn_data: &TransactionMetadata,
+        cleanup_args: Vec<Vec<u8>>,
+    ) -> Result<ChangeSetExt, VMStatus> {
+        // Charge gas for writeset before we do cleanup. This ensures we don't charge gas for
+        // cleanup writeset changes, which is consistent with outer-level success cleanup
+        // flow. We also wouldn't need to worry that we run out of gas when doing cleanup.
+        let inner_function_change_set_ext = session
+            .finish(&mut (), gas_meter.change_set_configs())
+            .map_err(|e| e.into_vm_status())?;
+        gas_meter.charge_write_set_gas_for_io(inner_function_change_set_ext.write_set().iter())?;
+        gas_meter.charge_storage_fee(
+            inner_function_change_set_ext.write_set().iter(),
+            inner_function_change_set_ext.change_set().events(),
+            txn_data.transaction_size,
+            txn_data.gas_unit_price,
+        )?;
```

Comment:
this same chunk of code seems to be in a few places

Reply:
Yep unfortunately. But they need to be done in different places in the entry function vs multisig tx execution flows. I'ev found a good way to deduplicate them yet.